File: petclinic-rest/src/main/java/org/springframework/samples/petclinic/service/ReviewService.java
package org.springframework.samples.petclinic.service;

import lombok.RequiredArgsConstructor;
import org.owasp.html.HtmlPolicyBuilder;
import org.owasp.html.PolicyFactory;
import org.springframework.stereotype.Service;
import org.springframework.samples.petclinic.mapper.ReviewMapper;
import org.springframework.samples.petclinic.model.Review;
import org.springframework.samples.petclinic.repository.ReviewRepository;
import org.springframework.samples.petclinic.rest.dto.ReviewStatsDto;

import java.util.List;

/**
 * Service for managing veterinarian reviews with XSS protection.
 */
@Service
@RequiredArgsConstructor
public class ReviewService {

    private final ReviewRepository reviewRepository;
    private final ReviewMapper reviewMapper;

    // OWASP HTML Sanitizer configured to strip all HTML tags (plain text only policy)
    private static final PolicyFactory PLAIN_TEXT_ONLY = new HtmlPolicyBuilder().toFactory();

    /**
     * Sanitizes feedback text using OWASP HTML Sanitizer to remove all HTML tags.
     * This provides XSS protection by stripping any potentially malicious content.
     *
     * @param feedback the raw feedback text from user input
     * @return sanitized plain text with all HTML tags removed
     */
    public String sanitizeFeedback(String feedback) {
        if (feedback == null || feedback.isEmpty()) {
            return feedback;
        }

        // Use OWASP HTML Sanitizer to strip all HTML tags and prevent XSS
        // This library is battle-tested and maintained by security experts
        String sanitized = PLAIN_TEXT_ONLY.sanitize(feedback);

        // Trim whitespace that may result from tag removal
        return sanitized.trim();
    }

    /**
     * Sanitizes the review feedback and saves it to the database.
     *
     * @param review the review to sanitize and save
     * @return the saved review with sanitized feedback
     */
    public Review sanitizeAndSave(Review review) {
        if (review.getFeedback() != null) {
            String sanitized = sanitizeFeedback(review.getFeedback());
            review.setFeedback(sanitized);
        }
        reviewRepository.save(review);
        return review;
    }

    /**
     * Retrieves all reviews for a specific veterinarian, sorted by creation date (most recent first).
     *
     * @param vetId the ID of the veterinarian
     * @return list of reviews sorted by createdAt descending
     */
    public List<Review> getReviewsByVetId(int vetId) {
        return reviewRepository.findByVetIdOrderByCreatedAtDesc(vetId);
    }

    /**
     * Calculates and retrieves aggregated review statistics for a veterinarian.
     *
     * @param vetId the ID of the veterinarian
     * @return review statistics including average rating and total count
     */
    public ReviewStatsDto getReviewStats(int vetId) {
        Double avgRating = reviewRepository.findAverageRatingByVetId(vetId);
        Long totalReviews = reviewRepository.countByVetId(vetId);

        ReviewStatsDto stats = new ReviewStatsDto();
        stats.setVetId(vetId);
        // Handle null values from aggregate queries when no reviews exist
        stats.setAverageRating(avgRating != null ? Math.round(avgRating * 10.0) / 10.0 : 0.0);
        stats.setTotalReviews(totalReviews != null ? totalReviews.intValue() : 0);

        return stats;
    }

    /**
     * Deletes a review from the database.
     *
     * @param review the review to delete
     */
    public void deleteReview(Review review) {
        reviewRepository.delete(review);
    }


    /**
     * Retrieves aggregated review statistics for all veterinarians in a single query.
     * This is more efficient than calling getReviewStats() for each vet individually.
     *
     * @return map of vet ID to review statistics
     */
    public java.util.Map<Integer, ReviewStatsDto> getAllReviewStats() {
        // Get all vets with reviews
        List<Object[]> results = reviewRepository.findAllReviewStats();

        java.util.Map<Integer, ReviewStatsDto> statsMap = new java.util.HashMap<>();

        for (Object[] result : results) {
            Integer vetId = (Integer) result[0];
            Double avgRating = (Double) result[1];
            Long totalReviews = (Long) result[2];

            ReviewStatsDto stats = new ReviewStatsDto();
            stats.setVetId(vetId);
            stats.setAverageRating(avgRating != null ? Math.round(avgRating * 10.0) / 10.0 : 0.0);
            stats.setTotalReviews(totalReviews != null ? totalReviews.intValue() : 0);

            statsMap.put(vetId, stats);
        }

        return statsMap;
    }

}

File: petclinic-angular/src/app/vets/review.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { catchError } from 'rxjs/operators';
import { environment } from '../../environments/environment';
import { HandleError, HttpErrorHandler } from '../error.service';
import { Review, ReviewSubmission, ReviewStats } from './review.model';

/**
 * Service for managing veterinarian reviews via HTTP API.
 * Handles CRUD operations for reviews and retrieves review statistics.
 */
@Injectable()
export class ReviewService {
  private readonly handlerError: HandleError;

  constructor(
    private http: HttpClient,
    private httpErrorHandler: HttpErrorHandler
  ) {
    this.handlerError = httpErrorHandler.createHandleError('ReviewService');
  }

  /**
   * Retrieves all reviews for a specific veterinarian.
   * Reviews are returned in reverse chronological order (most recent first).
   *
   * @param vetId - The ID of the veterinarian
   * @returns Observable of Review array
   */
  getReviewsByVetId(vetId: number): Observable<Review[]> {
    const url = `${environment.REST_API_URL}vets/${vetId}/reviews`;
    return this.http
      .get<Review[]>(url)
      .pipe(catchError(this.handlerError('getReviewsByVetId', [])));
  }

  /**
   * Retrieves aggregated statistics for a veterinarian's reviews.
   * Includes average rating, total count, and most recent review.
   *
   * @param vetId - The ID of the veterinarian
   * @returns Observable of ReviewStats
   */
  getReviewStats(vetId: number): Observable<ReviewStats> {
    const url = `${environment.REST_API_URL}vets/${vetId}/reviews/stats`;
    const emptyStats: ReviewStats = {
      vetId,
      averageRating: 0,
      totalReviews: 0
    };
    return this.http
      .get<ReviewStats>(url)
      .pipe(catchError(this.handlerError('getReviewStats', emptyStats)));
  }

  /**
   * Submits a new review for a veterinarian.
   * The review will be sanitized on the backend before storage.
   *
   * @param submission - The review data to submit
   * @returns Observable of the created Review with ID and timestamp
   */
  submitReview(submission: ReviewSubmission): Observable<Review> {
    const url = `${environment.REST_API_URL}vets/${submission.vetId}/reviews`;
    return this.http
      .post<Review>(url, {
        rating: submission.rating,
        feedback: submission.feedback
      })
      .pipe(catchError(this.handlerError('submitReview', {} as Review)));
  }

  /**
   * Deletes a review by ID.
   * This operation is typically restricted to administrators.
   *
   * @param reviewId - The ID of the review to delete
   * @returns Observable of void
   */
  deleteReview(reviewId: number): Observable<void> {
    const url = `${environment.REST_API_URL}reviews/${reviewId}`;
    return this.http
      .delete<void>(url)
      .pipe(catchError(this.handlerError('deleteReview', undefined)));
  }

  /**
   * Retrieves aggregated statistics for all veterinarians' reviews in a single request.
   * This is more efficient than calling getReviewStats() for each vet individually.
   *
   * @returns Observable of a map from vet ID to ReviewStats
   */
  getAllReviewStats(): Observable<{ [vetId: number]: ReviewStats }> {
    const url = `${environment.REST_API_URL}reviews/stats`;
    return this.http
      .get<{ [vetId: number]: ReviewStats }>(url)
      .pipe(catchError(this.handlerError('getAllReviewStats', {})));
  }
}
